coverage/hare_coverage.py

```diff
@@ -0,0 +1,319 @@
+        """
+        Initialize coverage params.
+        :param cov_type: coverage type xml or html
+        :param source_test_map: source module to test module map
+        """
+        self.cov_type: str = "html" if cov_type.lower() == "html" else "xml"
+        if source_test_map is None:
+            self.src_tst_map: Optional[Dict[str, str]] = self.source_test_map
+        else:
+            self.src_tst_map: Optional[Dict[str, str]] = source_test_map
+        self.report_dir = report_dir
+        self.report = self.report_dir + sep
+        self.report += "html_report" if self.cov_type == "html" else "coverage.xml"
+        self.utils = Utility()
+        logging.info("Running Python Coverage Tool.")
+
+    def cleanup(self) -> None:
+        """
+        Cleaning up old coverage reports.
+        :return: None
+        """
+        logging.info("Cleaning up the report.")
+        try:
+            if self.utils.is_exist(self.report_dir) is True:
+                self.utils.run_cmd(f"rm -rf {self.report_dir}")
+        except Exception as e:
+            logging.error(f"Error {e} while cleanup.")
+
+    def run_ut(self) -> None:
+        """
+        Execute unit testcases.
+        :return: None
+        """
+        logging.info("Running UT's.")
+        dir_scanner = DirectoryScanner("test")
+        for test_dir_path in dir_scanner.search_dir(getcwd()):
+            self.utils.run_cmd(f"pytest {test_dir_path}")
+
+    def run_coverage(self) -> None:
+        """
+        Running coverage tool over UTs.
+        :return: None
+        """
+        logging.info("Running Coverage.")
+        command_line: str
+        for index, (src, tst) in enumerate(self.src_tst_map.items()):
+            if index == 0:
+                command_line: str = f"pytest --cov={src} --cov-report={self.cov_type}:{self.report} {tst}"
+            else:
+                command_line: str = f"pytest --cov={src} --cov-report={self.cov_type}:{self.report} --cov-append {tst}"
+            self.utils.run_cmd(command_line)
+
+    def check_report(self) -> None:
+        """
+        Validate the report generated.
+        :return: None
+        """
+        logging.info("Checking report.")
+        if self.utils.is_exist(self.report) is True:
+            logging.info(f"Python Code Coverage Report generated successfully at {self.report!a}.")
+            print(f"Python Code Coverage Report generated successfully at {self.report!a}.")
+        else:
+            logging.error("Failed to generate the Coverage report.")
+            raise FailedToGenerateReport("Failed to generate the Coverage report.")
+
+
+class CCoverage(Strategy):
+
+    def cleanup(self) -> None:
+        """
+        Cleaning up old coverage reports.
+        :return: None
+        """
+        pass
+
+    def run_ut(self) -> None:
+        """
+        Execute unit testcases.
+        :return: None
+        """
+        pass
+
+    def run_coverage(self) -> None:

```

Comment:
May be we can use gcov for c code coverage.

Reply:
As we discuss into the DSU, for C code coverage we need to debug the setuptools.setup.py. Once that done we were able to complete the POC and then we can add code for that.